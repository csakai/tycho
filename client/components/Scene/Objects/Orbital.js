import THREE from 'three';
import Constants from 'constants';
import Math2 from 'engine/math2';
import Mesh from '../Utils/Mesh';
import Ellipse from '../Utils/Ellipse';
import Rings from './Rings';

export default class Orbital {

  /**
   * @param  {Object} data
   */
  constructor(data, parent) {
    this.updateQueue = [];
    this.data = data;
    this.setUp();
    this.renderChildren(parent);

    if(parent) {
      parent.add(this.getOrbital());
    }
  }

  /**
   * Renders the stage props that are part of this orbital
   */
  setUp = () => {
    this.ellipse = new Ellipse(this.data);
    this.mesh = new Mesh(this.data);
    
    if (this.data.ring) {
      this.mesh.body.add(new Rings(this.data.ring));
    }
  }

  /**
   * Renders all instances of children from within the queue.
   */
  renderChildren = () => {
    let children = this.data.children;

    if(children) {
      children.forEach(child => {
        let orbital = new Orbital(child, this.mesh);
        this.updateQueue.push(orbital.updatePosition);
      });
    }
  }

  /**
   * Light up orbital path
   */
  illuminatePath = () => {
    this.label.illuminated = true;
    this.ellipse.setPathBrightness(0.4);
  }

  /**
   * Darken orbital path
   */
  darkenPath = () => {
    this.label.illuminated = false;
    this.ellipse.setPathBrightness(1);
  }

  /**
   * Renders the orbital plane containing props
   * @return {Object3D}
   */
  getOrbitalPlane = () => {
    let orbitalPlane = new THREE.Object3D();

    orbitalPlane.add(this.ellipse);
    orbitalPlane.add(this.mesh);

    return orbitalPlane;
  }

  /**
   * Returns the mesh reference plane containing relevant props.
   * @param  {Object}   data
   * @return {Object3D}
   */
  getOrbital = () => {
    let referencePlane = new THREE.Object3D();
    let orbitalPlane   = this.getOrbitalPlane(this.data);

    referencePlane.add(orbitalPlane);
    this.setPlanarRotations(orbitalPlane, referencePlane);
    this.setAxialTilt();

    return referencePlane;
  }

  /**
   * Sets the y-order rotation of the mesh's body instance.
   * @param {Number} axialTilt
   */
  setAxialTilt = () => {
    this.rotateObject('x', this.mesh.body, this.data.axialTilt);
  }

  /**
   * Sets (x,z)-order rotations on subscene planes.
   * @param {[Object3D} orbitalPlane
   * @param {Object3D}  referencePlane
   */
  setPlanarRotations = (orbitalPlane, referencePlane) => {
    const {inclination, longAscNode, argPeriapsis} = this.data;

    this.rotateObject('x', referencePlane, inclination);
    this.rotateObject('z', referencePlane, longAscNode);
    this.rotateObject('z', orbitalPlane, argPeriapsis);
  }

  /**
   * Rotates an scene object.
   * @param  {String}   coordinate x, y, or z
   * @param  {Object3D} object     object to rotate
   * @param  {Number}   rotation   in degrees
   */
  rotateObject = (coordinate, object, rotation) => {
    if(isNaN(object[coordinate])) {
      object.rotation[coordinate] = 0;
    }
    object.rotation[coordinate] += Math2.toRadians(rotation);
  }

  /**
   * Update real-time attributes (velocity and position).
   * @param  {Number}  time UNIX time
   */
  updatePosition = (time) => {
    const pos = this.ellipse.getPosition(time, this.data.periapses);

    this.updateQueue.forEach((update) => update(time));
    this.mesh.updatePosition(time, pos);
  }
}
