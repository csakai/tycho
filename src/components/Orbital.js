import React from 'react';
import PropTypes from 'prop-types';
import Body from './Body';
import ReactAnimationFrame from 'react-animation-frame';

export class Orbital extends React.Component {

  static propTypes = {
    updateScreenPosition: PropTypes.func.isRequired,
    eclipticGroupRotation: PropTypes.object.isRequired,
    orbitalGroupRotation: PropTypes.object.isRequired,
    pathVertices: PropTypes.array.isRequired,
    bodyPosition: PropTypes.object.isRequired,
    bodyRotation: PropTypes.object.isRequired,
    bodyRadius: PropTypes.number.isRequired,
    pathOpacity: PropTypes.number,
    id: PropTypes.string.isRequired
  }

  onAnimationFrame = () => {
    this.props.updateScreenPosition(this.refs.body);
  }

  render() {
    return (
      <group rotation={this.props.eclipticGroupRotation}>
        <group rotation={this.props.orbitalGroupRotation} position={this.props.bodyPosition}>
          <Body
            rotation={this.props.bodyRotation}
            radius={this.props.bodyRadius}
            ref="body"
          />
          {this.props.children}
        </group>
        <line>
          <lineBasicMaterial color={0x0000ff} opacity={this.props.pathOpacity} />
          <geometry vertices={this.props.pathVertices} />
        </line>
      </group>
    );
  }
}

export default ReactAnimationFrame(Orbital);
