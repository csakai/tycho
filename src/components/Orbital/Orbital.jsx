import React from 'react';
import PropTypes from 'prop-types';
import Body from '../Body';
import Label from './Label';

export default class Orbital extends React.Component {

  static propTypes = {
    eclipticGroupRotation: PropTypes.object.isRequired,
    orbitalGroupRotation: PropTypes.object.isRequired,
    pathVertices: PropTypes.array.isRequired,
    bodyPosition: PropTypes.object.isRequired,
    bodyRotation: PropTypes.object.isRequired,
    bodyRadius: PropTypes.number.isRequired,
    atmosphere: PropTypes.number,
    pathOpacity: PropTypes.number,
    scaleLastUpdate: PropTypes.number,
    maps: PropTypes.array,
    targetId: PropTypes.string,
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    action: PropTypes.object.isRequired
  }

  render() {
    return (
      <group ref="ecliptic" rotation={this.props.eclipticGroupRotation}>
        <group ref="orbital" rotation={this.props.orbitalGroupRotation}>
          
          <group 
            position={this.props.bodyPosition}
            name={this.props.id}>
            <Body
              rotation={this.props.bodyRotation}
              radius={this.props.bodyRadius}
              maps={this.props.maps}
            />
            <Label {...this.props} />
            {this.props.children}
          </group>

          <line key={`path-${this.props.id}-${this.props.scaleLastUpdate}`}>
            <lineBasicMaterial
              transparent={true}
              color={this.props.atmosphere}
              opacity={this.props.pathOpacity}
            />
            <geometry
              vertices={this.props.pathVertices}
            />
          </line>
        </group>
      </group>
    );
  }
}
