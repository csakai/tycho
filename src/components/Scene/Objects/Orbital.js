import React from 'react';
import ReactAnimationFrame from 'react-animation-frame';
import PropTypes from 'prop-types';
import * as THREE from 'three';

import Ellipse from '../Utils/Ellipse';
import Mesh from '../Utils/Mesh';
import Math2 from '../../../engine/math2';

class Orbital extends React.Component {

  static propTypes = {
    inclination: PropTypes.number.isRequired,
    longAscNode: PropTypes.number.isRequired,
    argPeriapsis: PropTypes.number.isRequired,
    arcRotate: PropTypes.number.isRequired,
    radius: PropTypes.number.isRequired,
    axialTilt: PropTypes.number.isRequired,
    time: PropTypes.number,
    odd: PropTypes.bool
  }

  constructor(props, context) {
    super(props, context);

    this.state = {};
    this.ellipse = new Ellipse(this.props);
  }

  toEuler = ({x, y, z}) => { // TODO: move to math lib
    const toRad = (val) => val ? Math2.toRadians(val) : 0;
    return new THREE.Euler(toRad(x), toRad(y), toRad(z));
  }

  getEclipticGroupRotation = ({inclination, longitudeOfAscendingNode}) => {
    return this.toEuler({
      x: inclination - (this.props.odd ? 0 : 90),
      z: longitudeOfAscendingNode
    });
  }

  getOrbitalGroupRotation = ({argumentOfPeriapsis}) => {
    return this.toEuler({
      z: argumentOfPeriapsis
    });
  }

  getBodyRotation = () => {
    return this.toEuler({
      x: 90,
      y: Math2.arcSecToDeg(this.props.time, this.props.arcRotate)*2
    });
  }

  getBodyPosition = () => {
    return this.ellipse.getPosition(this.props.time, this.props.periapses);
  }

  onAnimationFrame = () => {
    this.setState({
      rotation: this.getBodyRotation(),
      position: this.getBodyPosition()
    });
  }

  renderBody = () => {
    // TODO: rename Mesh to Body
    return (
      <Mesh
        rotation={this.state.rotation}
        axialTilt={this.props.axialTilt} 
        radius={this.props.radius}
      />
    );
  }

  renderLine = () => {
    return (
      <line>
        <lineBasicMaterial color={0x0000ff} />
        <geometry vertices={this.ellipse.geometry.vertices} />
      </line>
    );
  }

  render() {
    return (
      <group rotation={this.getEclipticGroupRotation(this.props)}>
        <group rotation={this.getOrbitalGroupRotation(this.props)} position={this.state.position}>
          {this.renderBody()}
          {this.props.children}
        </group>
        {this.renderLine()}
      </group>
    );
  }
}

export default ReactAnimationFrame(Orbital);
