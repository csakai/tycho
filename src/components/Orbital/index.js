import ReactAnimationFrame from 'react-animation-frame';
import Orbital from './Orbital';

export default Orbital;//ReactAnimationFrame(Orbital);
