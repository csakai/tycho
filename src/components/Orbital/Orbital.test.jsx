import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Orbital from './Orbital';

describe('Orbital Component', () => {
  let component, orbital, updatePosition;

  beforeEach(() => {
    const position = {};
    const rotation = {};

    updatePosition = jest.fn();

    component = shallow(<Orbital
      updatePosition={updatePosition}
      eclipticGroupRotation={rotation}
      orbitalGroupRotation={rotation}
      pathVertices={[]}
      bodyPosition={position}
      bodyRotation={rotation}
      bodyRadius={1}
      id="testOrbital"
      text="Test Orbital"
      action={{}}
    />);

    orbital = component.instance();
  });

  describe('render()', () => {
    it('should render the orbital successfully', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
