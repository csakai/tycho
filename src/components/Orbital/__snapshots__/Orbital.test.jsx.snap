// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orbital Component render() should render the orbital successfully 1`] = `
<group
  rotation={Object {}}
>
  <group
    rotation={Object {}}
  >
    <group
      name="testOrbital"
      position={Object {}}
    >
      <Body
        radius={1}
        rotation={Object {}}
      />
    </group>
    <line
      key="path-testOrbital-undefined"
    >
      <lineBasicMaterial
        transparent={true}
      />
      <geometry
        vertices={Array []}
      />
    </line>
  </group>
</group>
`;
